pasture-algorithms/src/normal_estimation.rs

```diff
@@ -0,0 +1,551 @@
+        for point in nearest_points {
+            k_nn_buffer.push_point(*point);
+        }
+
+        // coordinates of the surface normal
+        let mut n_x: f64 = 0.0;
+        let mut n_y: f64 = 0.0;
+        let mut n_z: f64 = 0.0;
+
+        // curvature
+        let mut curvature: f64 = 0.0;
+
+        normal_estimation(&k_nn_buffer, &mut n_x, &mut n_y, &mut n_z, &mut curvature);
+
+        // solution vector
+        points_with_normals_curvature.push((point, n_x, n_y, n_z, curvature));
+    }
+
+    return points_with_normals_curvature;
+}
+
+/// checks whether a given point cloud has points with coordinates that are Not a Number
+fn is_dense<T: PointBuffer>(point_cloud: &T) -> bool {
+    for point in point_cloud.iter_attribute::<Vector3<f64>>(&POSITION_3D) {
+        if point.x.is_nan() || point.y.is_nan() || point.z.is_nan() {
+            return false;
+        }
+    }
+    return true;
+}
+
+/// checks whether a given point has finite coordinates
+fn is_finite(point: &Vector3<f64>) -> bool {
+    if !(point.x.is_finite() || point.y.is_finite() || point.z.is_finite()) {
+        return false;
+    }
+    return true;
+}
+
+/// computes the centroid for a given point cloud
+/// the centroid is the point that has the same distance to all other points in the point cloud
+fn compute_centroid<T: PointBuffer>(point_cloud: &T, centroid: &mut Vector4<f64>) {
+    if point_cloud.is_empty() {
+        panic!("The point cloud is empty!");
+    }
+
+    centroid.set_zero();
+    let mut temp_centroid = vec![0.0; 3];
+
+    if is_dense(point_cloud) {
+        // add all points up
+        for point in point_cloud.iter_attribute::<Vector3<f64>>(&POSITION_3D) {
+            temp_centroid[0] += point.x;
+            temp_centroid[1] += point.y;
+            temp_centroid[2] += point.z;
+        }
+
+        //normalize over all points
+        centroid[0] = temp_centroid[0] / point_cloud.len() as f64;
+        centroid[1] = temp_centroid[1] / point_cloud.len() as f64;
+        centroid[2] = temp_centroid[2] / point_cloud.len() as f64;
+        centroid[3] = 1.0;
+    } else {
+        let mut points_in_cloud = 0;
+        for point in point_cloud.iter_attribute::<Vector3<f64>>(&POSITION_3D) {
+            if is_finite(&point) {
+                // add all points up
+                temp_centroid[0] += point.x;
+                temp_centroid[1] += point.y;
+                temp_centroid[2] += point.z;
+                points_in_cloud = points_in_cloud + 1;
+            }
+        }
+
+        // normalize over all points
+        centroid[0] = temp_centroid[0] / points_in_cloud as f64;
+        centroid[1] = temp_centroid[1] / points_in_cloud as f64;
+        centroid[2] = temp_centroid[2] / points_in_cloud as f64;
+        centroid[3] = 1.0;
+    }
+}
+
+/// compute the covariance matrix for a given point cloud which is a measure of spread out the points are
+fn compute_covarianz_matrix<T: PointBuffer>(
```

Comment:
`covariance`, not `covarianz`.
Again, why do you pass output parameters in instead of returning them? In Rust you can just return a tuple: `-> (usize, DMatrix<f64>)`